Скачайте вот этот текст -  https://github.com/mannefedov/compling_nlp_hse_course/blob/master/data/zhivago.txt

### Задание 1 (2 балла)
Отчистите текст от мусора (тэгов, хешей и тп) с помощью регулярных выражений. Постарайтесь убрать весь мусор, но если что-то удалить не получается, то не мучайтесь. Главное, чтобы мусор не проявлялся в результатах следующих заданий.

In [1]:
#Скачайте вот этот текст - https://github.com/mannefedov/compling_nlp_hse_course/blob/master/data/zhivago.txt
import requests
url = 'https://raw.githubusercontent.com/mannefedov/compling_nlp_hse_course/master/data/zhivago.txt'
text = requests.get(url).text
text[:1000]

'<author><first-name>Борис</first-name><middle-name>Леонидович</middle-name><last-name>Пастернак</last-name></author>\n<book-title>Доктор Живаго</book-title>\n<annotation><p>«Доктор Живаго» - итоговое произведение Бориса Пастернака, книга всей его жизни. Этот роман принес его автору мировую известность и Нобелевскую премию, присуждение которой обернулось для поэта оголтелой политической травлей, обвинениями в «измене Родине» и в результате стоило ему жизни.</p>\n<p>«Доктор Живаго» - роман, сама ткань которого убедительнее свидетельствует о чуде, чем все размышления доктора и обобщения автора. Человек, который так пишет, бесконечно много пережил и передумал, и главные его чувства на свете  - восхищенное умиление и слезное сострадание; конечно, есть в его мире место и презрению, и холодному отстранению  - но не в них суть. Роман Пастернака  - оплакивание прежних заблуждений и их жертв; те, кто не разделяет молитвенного восторга перед миром, достойны прежде всего жалости. Перечитывать «До

In [2]:
#Отчистите текст от мусора (тэгов, хешей и тп) с помощью регулярных выражений.
import re
text = re.sub(u'\xa0', ' ', text)
text = re.sub(r'<[^>]+>', ' ', text)
text = re.sub(r'\n', ' ', text)
text = re.sub('  +', ' ', text)
text[:1000]

' Борис Леонидович Пастернак Доктор Живаго «Доктор Живаго» - итоговое произведение Бориса Пастернака, книга всей его жизни. Этот роман принес его автору мировую известность и Нобелевскую премию, присуждение которой обернулось для поэта оголтелой политической травлей, обвинениями в «измене Родине» и в результате стоило ему жизни. «Доктор Живаго» - роман, сама ткань которого убедительнее свидетельствует о чуде, чем все размышления доктора и обобщения автора. Человек, который так пишет, бесконечно много пережил и передумал, и главные его чувства на свете - восхищенное умиление и слезное сострадание; конечно, есть в его мире место и презрению, и холодному отстранению - но не в них суть. Роман Пастернака - оплакивание прежних заблуждений и их жертв; те, кто не разделяет молитвенного восторга перед миром, достойны прежде всего жалости. Перечитывать «Доктора Живаго» стоит именно тогда, когда кажется, что жить не стоит. Тогда десять строк из этого романа могут сделать то же, что делает любовь 

### Задание 2. (2 балла) 
Приведите очищенный текст к нижнему регистру, удалите все знаки пунктуации, разделите на предложения библиотекой rusenttokenize, токенизируйте библиотекой razdel_tokenize. 
Ответьте на следующие вопросы:
1) есть ли в тексте повторяющиеся корректные предложения? если да то какие? (если находится мусор то вернитесь к заданию 1 и постарайтесь избавиться от него)
2) какой самый частотный токен в тексте длиннее 6 символов?


In [3]:
!pip install rusenttokenize
!pip install razdel
from collections import Counter
from string import punctuation
from rusenttokenize import ru_sent_tokenize
from razdel import tokenize as razdel_tokenize

#Приведите очищенный текст к нижнему регистру
text = text.lower()

#Удалите все знаки пунктуации, токенизируйте библиотекой razdel_tokenize
punctuation = f'{punctuation}«»–...'
tokens = [token.text.strip(punctuation) for token in razdel_tokenize(text) 
          if token.text not in punctuation]
tokens = [token for token in tokens if token]
print(tokens[:15])

['борис', 'леонидович', 'пастернак', 'доктор', 'живаго', 'доктор', 'живаго', 'итоговое', 'произведение', 'бориса', 'пастернака', 'книга', 'всей', 'его', 'жизни']


In [4]:
#Разделите на предложения библиотекой rusenttokenize
sentences = ru_sent_tokenize(text)

#Есть ли в тексте повторяющиеся корректные предложения? Если да, то какие?
def sorted_dict(list):
    counter = dict(Counter(list)).items()
    result = sorted(counter, key=lambda x: x[1], reverse=True)
    return result

[sent for sent in sorted_dict(sentences) if sent[1] > 1]

[('свеча горела на столе, свеча горела.', 3),
 ('парило.', 2),
 ('странно.', 2),
 ('толпа росла.', 2),
 ('да.', 2),
 ('он открыл глаза.', 2)]

In [5]:
[word for word in sorted_dict(tokens) if len(word[0]) > 6][0]

('андреевич', 289)

### Задание 3. (2 балла)

Сделайте стемминг и найдите по несколько частотных примеров на каждый из видов ошибок:
1) два разных слова ошибочно свелись к одинаковой основе
3) слово не изменилось после стемминга (слово должно быть русским и длиннее 4 символов)

***не делайте это задание полностью вручную, напишите правила, которые отловят потенциальные ошибки и выберите из них***


In [6]:
from nltk.stem.snowball import SnowballStemmer
stemmer = SnowballStemmer('russian')
words = [token.text for token in list(razdel_tokenize(text))]

stems = {}
stem_counter = Counter(words)

for word in words:
    stems[word] = (stem_counter[stemmer.stem(word)],stemmer.stem(word))

stem_dict = {} 
for key, value in stems.items(): 
    if value not in stem_dict: 
        stem_dict[value] = [key]
    else:
        stem_dict[value].append(key) 
        
stem_list = sorted(stem_dict.items(), reverse = True)

In [7]:
stem_list[:20]

[((18653, ','), [',']),
 ((12860, '.'), ['.']),
 ((6475, 'и'), ['и']),
 ((4748, 'в'), ['в']),
 ((2933, 'не'), ['не', 'ней', 'нее', 'нею']),
 ((2663, 'на'), ['на']),
 ((2421, '–'), ['–']),
 ((1929, 'с'), ['с']),
 ((1577, 'он'), ['они', 'он', 'оно', 'она', 'оной']),
 ((1489, 'что'), ['что']),
 ((1288, 'как'),
  ['как',
   'какие',
   'какой',
   'какая',
   'каком',
   'каким',
   'какую',
   'какое',
   'каких',
   'какою',
   'какому',
   'каков',
   'какими',
   'какого',
   'кака',
   'каки']),
 ((1196, '?'), ['?']),
 ((1190, 'я'), ['я']),
 ((1037, 'к'), ['к']),
 ((990, 'а'), ['а', 'ай']),
 ((964, 'по'), ['по', 'поила', 'поили']),
 ((796, 'из'), ['из', 'изо']),
 ((761, 'все'), ['всей', 'все', 'всею', 'всё']),
 ((729, 'у'), ['у', 'уа']),
 ((723, 'но'), ['но'])]

Слова, которые ошибочно свелись к одной основе:

((2933, 'не'), ['не', 'ней', 'нее', 'нею'])  
((1577, 'он'), ['они', 'он', 'оно', 'она', 'оной'])  
((1288, 'как'), ['как', 'какие', 'какой', 'какая', 'каком', 'каким', 'какую', 'какое', 'каких', 'какою', 'какому', 'каков', 'какими', 'какого', 'кака', 'каки'])  
((990, 'а'), ['а', 'ай'])  
((964, 'по'), ['по', 'поила', 'поили'])  
((761, 'все'), ['всей', 'все', 'всею', 'всё'])  
((729, 'у'), ['у', 'уа'])

In [8]:
#Слова, которые не изменились после стемминга:

stems = {}
for word in words:
    stems[stem_counter[word],word] = stemmer.stem(word)

unchanged_dict = {}
for key, value in stems.items():
    if key[1] == value and len(key[1])>4:
        unchanged_dict[key] = value

unchanged_list = sorted(unchanged_dict.items(), reverse = True)

In [9]:
unchanged_list[:10]

[((289, 'андреевич'), 'андреевич'),
 ((226, 'доктор'), 'доктор'),
 ((172, 'перед'), 'перед'),
 ((162, 'может'), 'может'),
 ((144, 'через'), 'через'),
 ((141, 'будет'), 'будет'),
 ((127, 'сейчас'), 'сейчас'),
 ((112, 'человек'), 'человек'),
 ((109, 'вдруг'), 'вдруг'),
 ((65, 'поезд'), 'поезд')]

Слова, которые не изменились после стемминга:

((289, 'андреевич'), 'андреевич')  
((226, 'доктор'), 'доктор')  
((172, 'перед'), 'перед')  
((162, 'может'), 'может')  
((144, 'через'), 'через')  
((141, 'будет'), 'будет')  
((127, 'сейчас'), 'сейчас')  
((112, 'человек'), 'человек')  
((109, 'вдруг'), 'вдруг')  
((65, 'поезд'), 'поезд')  

### Задание 4 (1 балл)
Проанализируйте список стоп-слов из нлтк (для русского). Какие ещё слова вы бы туда добавили? (5 слов будет достаточно, не забудьте аргументировать ваш выбор)


In [10]:
!pip install nltk
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

stops = stopwords.words('russian')

ruscorpora = ['и', 'в', 'не', 'на', 'я', 'быть', 'он', 'с', 'что', 'а', 'по', 'это', 'она', 'этот', 'к', 'но', 'они', 'мы', 'как', 'из', 'у', 'который', 'то', 'за', 'свой', 'что', 'весь', 'год', 'от', 'так', 'о', 'для', 'ты', 'же', 'все', 'тот', 'мочь', 'вы', 'человек', 'такой', 'его', 'сказать', 'только', 'или', 'ещё', 'бы', 'себя', 'один', 'как', 'уже', 'до', 'время', 'если', 'сам', 'когда', 'другой', 'вот', 'говорить', 'наш', 'мой', 'знать', 'стать', 'при', 'чтобы', 'дело', 'жизнь', 'кто', 'первый', 'очень', 'два', 'день', 'её', 'новый', 'рука', 'даже', 'во', 'со', 'раз', 'где', 'там', 'под', 'можно', 'ну', 'какой', 'после', 'их', 'работа', 'без', 'самый', 'потом', 'надо', 'хотеть', 'ли', 'слово', 'идти', 'большой', 'должен', 'место', 'иметь', 'ничто']

for word in ruscorpora:
    if word.replace('ё', 'е') not in stops:
        print(word)


это
который
свой
весь
год
мочь
человек
сказать
время
говорить
наш
знать
стать
дело
жизнь
первый
очень
день
новый
рука
работа
самый
хотеть
слово
идти
большой
должен
место
иметь
ничто


[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/datanikitin/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Ответ: все слова, не передающие большого смысла, из списка 1000 самых встречаемых согласно НКРЯ, которые еще не вошли в NLTK stopwords - это, который, свой, весь, наш, очень.

### Задание 5 (3 балла)

Предобработайте текст двумя способами:
1) лемматизируйте токены с помощью pymorphy2
2) лемматизируйте текст с помощью mystem3 

Ответьте на вопрос:
Что в данном случае лучше для лемматизации mystem или pymorphy?

Важно, чтобы ответы на вопросы были аргументированы (как минимум три аргумента).  Анализируйте результаты с языковой точки зрения. Скорость работы простота интерфейса не являются преимуществами или недостатками в рамках этого задания. Идеальный аргумент - библиотека x неправильно обрабатывает вот такое слово/класс слов. 

In [11]:
import pandas as pd
!pip install pymorphy2
!pip install pymystem3

import pymorphy2
from pymystem3 import Mystem

mystem = Mystem()
morph =  pymorphy2.MorphAnalyzer()

#не обрабатывал весь текст, чтобы не ждать mystem вечность
lemmas_pymorphy = [morph.parse(token)[0].normal_form for token in tokens[:200]]
lemmas_mystem = mystem.lemmatize(text[:2000])

lemmas_mystem_clean = []
for i in lemmas_mystem:
  if re.findall(r'\w+', i) != []:
    lemmas_mystem_clean.append(i)

for i in range(150):
  if lemmas_pymorphy[i] != lemmas_mystem_clean[i]:
    print(i, tokens[i], lemmas_pymorphy[i], lemmas_mystem_clean[i])

13 его он его
17 принес принести приносить
27 обернулось обернуться обертываться
49 убедительнее убедительный убедительно
54 все всё весь
66 пережил пережить переживать
68 передумал передумать передумывать
71 его он его
75 восхищенное восхитить восхищенный
78 слезное слёзный слезный
81 есть есть быть
83 его он его
95 суть быть суть
102 их они их
114 всего весь все
119 стоит стоить стоять
132 этого это этот


На этом датасете у меня показал себя лучше pymorphy.

Аргумент 1: mystem теряет информацию о букве ё (слезное слёзный слезный). Для некоторых слов это может быть критично.  
Аргумент 2: mystem не использует контекст, поэтому не может образовать правильно форму от "стоит" (стоит стоить стоять)  
Аргумент 3: mystem теряет информацию о виде глагола в отличие от pymorphy (передумал передумать передумывать)